### PR TITLE
Added TC for NRFU4.4 Multi agent routing protocol functionality

### PR DESCRIPTION
**Pull request by vitthalmagadum**
# Please include a summary of the changes

test_definition.yaml: Updated test def with testcase NRFU4.4
test_definition.yaml.j2: Updated J2 with testcase NRFU4.4
test_routing_multi_agent.py: Added python file for testcase NRFU4.4

# Any specific logic/part of code you need extra attention on

Running 'show ip route summary' command and verifying multi agent routing model protocol is configured and operational on the device

# Issue_ID: https://github.com/aristanetworks/vane/issues/433

# Type of change

- - [ ] Bug fix (non-breaking change which fixes an issue)
- - [ ] New feature (non-breaking change which adds functionality)
- - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- - [ ] This change requires a documentation update
- - [X] Other (NRFU Tests)

# Effort required on reviewers end

- - [X] Easy
- - [ ] Medium
- - [ ] Hard 

# Reports

Passed reports when multi agent enabled on device
[report_passed.docx](https://github.com/aristanetworks/vane/files/12483586/report_passed.docx)

Passed report using j2
[report_passed_j2.docx](https://github.com/aristanetworks/vane/files/12495441/report_passed_j2.docx)

Failed report when multi agent routing model protocol is not enabled(ribd mode enabled)
[report_failed.docx](https://github.com/aristanetworks/vane/files/12495458/report_failed.docx)

Failed reports when multi agent routing protocol is configured but not in operational state (tested using hardcode output)
[report_failed_not_operational_state.docx](https://github.com/aristanetworks/vane/files/12485747/report_failed_not_operational_state.docx)

Failed reports when routing details not found.(tested using hardcode output)
[report_assert_failed.docx](https://github.com/aristanetworks/vane/files/12485745/report_assert_failed.docx)

Html Reports
[html_reports.zip](https://github.com/aristanetworks/vane/files/12485743/html_reports.zip)

### CODE REVIEW

File: nrfu_tests/test_routing_multi_agent.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                tops.show_cmd,
+                route_summary,
+            )
+            self.output += (
+                f"On device {tops.dut_name}, Output of {tops.show_cmd} is:\n{route_summary}\n"
+            )
+            assert route_summary, "Routing model details are not found."
+
+            # Verify multi-agent is configured
+            tops.actual_output["multi_agent_routing_protocol"]["configured"] = (
```
Comment (hsisalgslab) [inline nrfu_tests/test_routing_multi_agent.py:63]:
```suggestion
            tops.actual_output["multi_agent_routing_protocol"]["configured_protocol_model"] = (
```
can we update like this ?

File: nrfu_tests/test_routing_multi_agent.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                configured_model in route_summary
+            )
+
+            # Verify multi-agent is configured
+            tops.actual_output["multi_agent_routing_protocol"]["operational"] = (
```
Comment (hsisalgslab) [inline nrfu_tests/test_routing_multi_agent.py:68]:
```suggestion
            tops.actual_output["multi_agent_routing_protocol"]["operational_protocol_model"] = (
```

File: nrfu_tests/test_routing_multi_agent.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        configured_model = "Configured routing protocol model: multi-agent"
+
+        # Forming output message if test result is pass
+        tops.output_msg = (
+            "Multi Agent Routing Model protocol is configured and operational on the device."
+        )
+
+        try:
+            """
+            TS: Running 'show ip route summary' command and Verifying Multi Agent Routing Model
```
Comment (hsisalgslab) [inline nrfu_tests/test_routing_multi_agent.py:47]:
```suggestion
            TS: Running 'show ip route summary' command and verifying multi agent routing model protocol is configured and operational on the device.
```

File: nrfu_tests/test_routing_multi_agent.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+        tops = tests_tools.TestOps(tests_definitions, TEST_SUITE, dut)
+        self.output = ""
+        tops.actual_output = {"multi_agent_routing_protocol": {}}
+        operating_model = "Operating routing protocol model: multi-agent"
+        configured_model = "Configured routing protocol model: multi-agent"
+
+        # Forming output message if test result is pass
+        tops.output_msg = (
+            "Multi Agent Routing Model protocol is configured and operational on the device."
```
Comment (hsisalgslab) [inline nrfu_tests/test_routing_multi_agent.py:42]:
```suggestion
            "Multi agent routing model protocol is configured and operational on the device."
```

File: nrfu_tests/test_routing_multi_agent.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            # Verify multi-agent is configured
+            tops.actual_output["multi_agent_routing_protocol"]["operational"] = (
+                operating_model in route_summary
+            )
+
+            # forming output message if test result is fail
+            if tops.actual_output != tops.expected_output:
+                tops.output_msg = f"\nOn device {tops.dut_name}:\n"
+                if not tops.actual_output["multi_agent_routing_protocol"]["operational"]:
+                    if tops.actual_output["multi_agent_routing_protocol"]["configured"]:
```
Comment (hsisalgslab) [inline nrfu_tests/test_routing_multi_agent.py:76]: add unit tests for following scenarios 
1) not is in operational state
2) Routing model details are not found.
3)Multi-agent is not configured.

File: nrfu_tests/test_definition.yaml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+- name: nrfu_tests
+  testcases:
```
Comment (geetanjalimanegslab) [inline nrfu_tests/test_definition.yaml:2]: Please follow the MR description format.

File: nrfu_tests/test_definition.yaml.j2 [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  testcases:
+  - name: test_routing_multi_agent
+    description: Test case for verification of multi-agent routing model functionality.
+    test_id: NRFU4.4
+    show_cmd: show ip route summary
+    expected_output:
+      multi_agent_routing_protocol:
+        configured: True
+        operational: True
+    test_criteria: Multi Agent Routing Model protocol should be configured and operational on the device.
```
Comment (geetanjalimanegslab) [inline nrfu_tests/test_definition.yaml.j2:11]:
```suggestion
    test_criteria: Multi agent routing model protocol should be configured and operational on the device.
```

File: nrfu_tests/test_routing_multi_agent.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
+# Copyright (c) 2023 Arista Networks, Inc.  All rights reserved.
+# Arista Networks, Inc. Confidential and Proprietary.
+
+"""Testcases for verification of multi-agent routing functionality"""
+
+import pytest
+from pyeapi.eapilib import EapiError
```
Comment (geetanjalimanegslab) [inline nrfu_tests/test_routing_multi_agent.py:7]: image

Any reason you have mentioned dut name here?

File: nrfu_tests/test_routing_multi_agent.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+                operating_model in route_summary
+            )
+
+            # forming output message if test result is fail
+            if tops.actual_output != tops.expected_output:
+                tops.output_msg = f"\nOn device {tops.dut_name}:\n"
+                if not tops.actual_output["multi_agent_routing_protocol"]["operational"]:
+                    if tops.actual_output["multi_agent_routing_protocol"]["configured"]:
+                        tops.output_msg += (
+                            "Multi agent routing protocol is configured but not is in operational"
```
Comment (geetanjalimanegslab) [inline nrfu_tests/test_routing_multi_agent.py:78]:
```suggestion
                            "Multi agent routing protocol is configured but it is not in the operational"
```

File: nrfu_tests/test_routing_multi_agent.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+                            " state."
+                        )
+                    if not tops.actual_output["multi_agent_routing_protocol"]["configured"]:
+                        tops.output_msg += (
+                            "Multi agent routing protocol is neither configured nor is in"
```
Comment (geetanjalimanegslab) [inline nrfu_tests/test_routing_multi_agent.py:83]:
```suggestion
                            "Multi agent routing protocol is neither configured nor in"
```

**Review by geetanjalimanegslab — CHANGES_REQUESTED** [review]

**Review by geetanjalimanegslab — APPROVED** [review]

**Review by hsisalgslab — APPROVED** [review]

File: nrfu_tests/test_routing_multi_agent.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+class MultiAgentRoutingProtocolTests:
+
+    """Testcases for verification of multi-agent routing functionality"""
+
+    dut_parameters = tests_tools.parametrize_duts(TEST_SUITE, test_defs, dut_objs)
+    test_duts = dut_parameters["test_routing_multi_agent"]["duts"]
+    test_ids = dut_parameters["test_routing_multi_agent"]["ids"]
+
+    @pytest.mark.parametrize("dut", test_duts, ids=test_ids)
+    def test_routing_multi_agent(self, dut, tests_definitions):
```
Comment (MaheshGSLAB) [inline nrfu_tests/test_routing_multi_agent.py:26]:
```suggestion
    def test_multi_agent_routing_protocol(self, dut, tests_definitions):
```

File: nrfu_tests/test_routing_multi_agent.py [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+        """
+        TD: Test case for verification of multi-agent routing model functionality
+        Args:
+          dut(dict): Details related to the switches
```
Comment (MaheshGSLAB) [inline nrfu_tests/test_routing_multi_agent.py:30]:
```suggestion
          dut(dict): Details related to particular DUT
```

File: nrfu_tests/test_routing_multi_agent.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+        tops.output_msg = (
+            "Multi agent routing model protocol is configured and operational on the device."
+        )
+
+        try:
+            """
+            TS: Running 'show ip route summary' command and verifying multi agent routing model
+            protocol is configured and operational on the device.
+            """
+            route_summary = dut["output"][tops.show_cmd]["text"]
```
Comment (MaheshGSLAB) [inline nrfu_tests/test_routing_multi_agent.py:50]: just want to know can we use {'cmd':'show ip route summary', 'revision': 3} to get the json output. Did you tried if yes then want is the observation

**Review by MaheshGSLAB — CHANGES_REQUESTED** [review]

**Review by MaheshGSLAB — APPROVED** [review]

**Review by kgrozis-arista — APPROVED** [review]

**Pull request merged**